static/js/services/contact-save.js

```diff
@@ -0,0 +1,162 @@
+        return _.map(repeated.child_data, function(child) {
+          child.parent = ExtractLineage(doc);
+          return prepare(child);
+        });
+      } else {
+        return [];
+      }
+    }
+
+    // Mutates the passed doc to attach prepared siblings, and returns all
+    // prepared siblings to be persisted
+    function prepareAndAttachSiblingDocs(schema, doc, original, siblings) {
+      if (!doc._id) {
+        throw new Error('doc passed must already be prepared with an _id');
+      }
+
+      var preparedSiblings = [];
+      var promiseChain = $q.resolve();
+
+      _.each(schema.fields, function(fieldConf, fieldName) {
+        var customType = fieldConf.type.match(/^(db|custom):(.*)$/);
+        if (customType) {
+          if(!doc[fieldName]) {
+            doc[fieldName] = {};
+          } else if(doc[fieldName] === 'NEW') {
+            var preparedSibling = prepare(siblings[fieldName]);
+            preparedSibling.type = customType[2];
+
+            var isChild = preparedSibling.parent === 'PARENT' ||
+                          (!preparedSibling.parent && fieldConf.parent === 'PARENT');
+
+            if (isChild) {
+              delete preparedSibling.parent;
+              // Cloning to avoid the circular reference we would make:
+              //   doc.fieldName.parent.fieldName.parent...
+              doc[fieldName] = _.clone(preparedSibling);
+              preparedSibling.parent = doc;
+            } else {
+              if (fieldName === 'parent' || fieldName === 'contact') {
+                doc[fieldName] = ExtractLineage(preparedSibling);
+              } else {
+                doc[fieldName] = preparedSibling;
+              }
+            }
+
+            preparedSiblings.push(preparedSibling);
+          } else if(original &&
+                    original[fieldName] &&
+                    original[fieldName]._id === doc[fieldName]) {
+
+            doc[fieldName] = original[fieldName];
+          } else {
+            promiseChain = promiseChain.then(function() {
+              return DB().get(doc[fieldName]).then(function(dbFieldValue) {
+                if (fieldName === 'parent' || fieldName === 'contact') {
+                  doc[fieldName] = ExtractLineage(dbFieldValue);
+                } else {
+                  doc[fieldName] = dbFieldValue;
+                }
+              });
+            });
+          }
+        }
+      });
+
+      return promiseChain.then(function() {
+        return preparedSiblings;
+      });
+    }
+
+    return function(schema, form, docId, type) {
+      return $q.resolve()
+        .then(function() {
+          if(docId) {
+            return DB().get(docId);
+          }
+          return null;
+        })
+        .then(function(original) {
+          var submitted = EnketoTranslation.contactRecordToJs(form.getDataStr());
+
+          if(original) {
+            submitted.doc = $.extend({}, original, submitted.doc);
```

Comment:
Use underscore not jquery for non-DOM related things.

Reply:
I didn't write this, it's just moved from the old service.  If you know exactly how to rewrite this please suggest :slightly_smiling_face:

Reply:
> Jquery does a deep copy of sub-objects, while underscore.js does a shallow copy.

If this is true, `underscore.extend()` probably wouldn't be suitable here.

Reply:
I think it's ok (and much faster) to do a shallow copy. Originally I thought that quote meant a deep merge but it appears that neither library do a deep merge: https://jsfiddle.net/m1o5rooe/

I think this code is cleaner and equivalent for our use:

```js
_.defaults(submitted.doc, original);
```

The only problem would be if we then modify properties in defaulted objects and then try and save the original doc but that seems unlikely.